### PR TITLE
Add module documentation for Mealy

### LINKED ISSUE #35: Library doesn't have sufficient module documentation
**Is your change request related to a problem? Please describe.**
As described in the [documentation section of the Library Guidelines](https://github.com/purescript-contrib/governance/blob/main/library-guidelines.md#documentation), Contributors libraries are expected to have mostly complete module documentation, preferably uploaded to Pursuit.

This library currently doesn't have sufficient module documentation -- many publicly-exported types and functions don't have any documentation comments.

**Describe the solution you'd like**
The library needs to have documentation comments on the majority (and preferably all) public types and functions. Once updated, we should tag a release and upload new module documentation to Pursuit so that folks can use it.

**Additional context**
See the [Governance repository](https://github.com/purescript-contrib/governance) for more information about requirements in the Contributors organization.

### PR DESCRIPTION
**Pull request by vladciobanu**
**Description of the change**
Fixes #35 . It's far from perfect, but perhaps it's a decent start?

Should I update the changelog if I'm not changing any code?

---

**Checklist:**

- [ ] Added the change to the changelog's "Unreleased" section with a link to this PR and your username
- [x] Linked any existing issues or proposals that this pull request should close
- [x] Updated or added relevant documentation in the README and/or documentation directory
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

**Comment by thomashoneyman** [comment]
Thanks! You can update the changelog if you would like, under the "Other improvements" section. Documentation updates are worthy of recognition in the changelog, I think, though it's not necessary if you don't want to.

**Comment by JordanMartinez** [comment]
I think doc updates should be included, not so much for the recognition, but more because those who didn't understand this library beforehand can be notified to look at it again and see whether it makes sense now.

**Comment by vladciobanu** [comment]
All right, will do so on both PR's.

File: src/Data/Machine/Mealy.purs [hunk 1]
```diff
@@ -47,62 +49,115 @@ import Data.Tuple (Tuple(..), fst, snd, swap)
...
 source src =  mealy $ \_ -> flip Emit (source src) <$> src
 
+-- | Sinks are 'terminator nodes' in machines. They allow for an
+-- | effectful computation to be executed on the inputs.
+-- |
+-- | For example, logging could be used as a sink:
+-- | ```purescript
+-- | take 10 $ source (pure 1) >>> sink logShow
+-- | ```
 sink :: forall f a. (Monad f) => (a -> f Unit) -> Sink f a
```
Comment (thomashoneyman) [inline src/Data/Machine/Mealy.purs:102]: As I look at these, this `Monad` constraint looks too powerful. Just a thought, but perhaps in a later PR we could relax this.

File: src/Data/Machine/Mealy.purs [hunk 2]
```diff
@@ -47,62 +49,115 @@ import Data.Tuple (Tuple(..), fst, snd, swap)
...
+-- | indefinitely.
+-- |
+-- | For example, generating ten instances of the value 1:
+-- | ```purescript
+-- | take 10 $ source (pure 1)
+-- | ```
 source :: forall f s. (Monad f) => f s -> Source f s
 source src =  mealy $ \_ -> flip Emit (source src) <$> src
 
+-- | Sinks are 'terminator nodes' in machines. They allow for an
```
Comment (thomashoneyman) [inline src/Data/Machine/Mealy.purs:95]: I think this part of the module comment might be better-suited for the `Sink` type (describing its purpose), and the comment for this function could focus on construction ("Construct a machine which executes an effectful computation on its inputs")

File: src/Data/Machine/Mealy.purs [hunk 3]
```diff
@@ -122,49 +178,69 @@ toUnfoldable s = unfoldr stepUnfold
...
                   go n                        = fromMaybe (a !! n) <> go (n + one)
               in  go zero
 
+-- | Creates a machine which wraps an effectful computation and ignores
+-- | its input.
 wrapEffect :: forall f s a. (Monad f) => f a -> MealyT f s a
 wrapEffect fa = MealyT $ const (flip Emit halt <$> fa)
 
 -- MonadLogic -- TODO: Create a purescript-logic package
+-- | Unwrap a machine such that its output is either `Nothign` in case
```
Comment (thomashoneyman) [inline src/Data/Machine/Mealy.purs:220]:
```suggestion
-- | Unwrap a machine such that its output is either `Nothing` in case
```

**Review by thomashoneyman — COMMENTED** [review]
I have a couple suggestions in here, but overall this is a _serious_ improvement and the comments you've written are concise and clear. Thank you for taking the time to add these. 👍

File: src/Data/Machine/Mealy.purs (hunk 1 above)
Comment (vladciobanu) [reply src/Data/Machine/Mealy.purs:102]: Yeah, I also noticed that but assumed it was a deliberate decision that I didn't understand...? If it's not, then I'll be happy to fix them.

**Comment by vladciobanu** [comment]
I think I addressed all review commends. I also added a line in the changelog.

File: src/Data/Machine/Mealy.purs (hunk 1 above)
Comment (thomashoneyman) [reply src/Data/Machine/Mealy.purs:103]: I think we ought to, if you’re up for it!

**Review by thomashoneyman — APPROVED** [review]

**Comment by thomashoneyman** [comment]
Excellent! Thanks so much.

**Pull request merged**